cli/popper/utils.py

```diff
@@ -104,19 +116,112 @@ def is_popperized(rootfolder):
     return os.path.isfile(config_filename)
 
 
+def setup_logging(level='INFO'):
+    root = logging.getLogger()
+    f = colorlog.ColoredFormatter('%(log_color)s%(levelname)s: %(message)s ',
+                                  log_colors=log_colors)
+    ch = logging.StreamHandler(sys.stdout)
+    ch.setFormatter(f)
+    root.addHandler(ch)
+    set_level(level)
+
+
+def set_level(level):
+    levels = {
+        'DEBUG': logging.DEBUG,
+        'INFO': logging.INFO,
+        'WARNING': logging.WARNING,
+        'ERROR': logging.ERROR,
+        'CRITICAL': logging.CRITICAL
+    }
+    root = logging.getLogger()
+    root.setLevel(levels[level])
+
+
+def remove_level():
+    handler = logging.getLogger().handlers[0]
+    f = colorlog.ColoredFormatter('%(log_color)s%(message)s ',
+                                  log_colors=log_colors)
+    handler.setFormatter(f)
+
+
+def add_level():
+    handler = logging.getLogger().handlers[0]
+    f = colorlog.ColoredFormatter('%(log_color)s%(levelname)s: %(message)s ',
+                                  log_colors=log_colors)
+    handler.setFormatter(f)
+
+
+def multiline(log):
+    """
+    A decorator for multiline log.
+    This solves a few problems for multiline log.
+    1. Without the decorator, a multiline log will have coloring for the first line
+    2. Only the first line should have level information
+        Eg. log('This is a multiline\nline') should print
+         LEVEL:
+         This is a multiline
+         line log
+        Note: LEVEL is printed on a separate line, so that log can retain indent info
+              This is ugly.
+              LEVEL:    Each line in
+                  this log is indented
+                  by four
+                  spaces
+    Note: 'INFO' level doesn't have color and level. So decorator is not required
+    """
+    def wrapper(msg):
+        # Check if it is a multiline log
+        msg_lines = msg.splitlines(True)
+        pos = 0
+        line_count = 0
+        for line in msg_lines:
+            if not line.isspace():
+                line_count += 1
+        if line_count > 1:
+            if msg_lines[0].isspace():
+                remove_level()
+                for index, line in enumerate(msg_lines):
+                    if line.isspace():
+                        log(line)
+                    else:
+                        pos = index
+                        break
+            if pos < len(msg_lines):
+                add_level()
+                log('')
+                remove_level()
+                rest = ''.join(msg_lines[index:])
+                log(rest+"\n")
+        else:
+            log(msg)
+    return wrapper
+
+
+@multiline
```

Comment:
since we are now using `logging`, it would be nice if we can use `log.fail`, `log.warn`, etc... directly on the code without having to wrap those calls. In the case of multiline output, we could leave that to the code that is invoking the logging call. This means that the code doing:

```python
log.debug('this is a multi-line output \n that can be put in multiple logging calls')
```

can be transformed to this:

```python
log.debug('this is a multi-line output')
log.debug(' .   that can be put in multiple')
log.debug(' .   logging calls')
```

in this way we don't need to worry and can make use of `logging` directly.

Reply:
In the case of `FAIL`, we can implement a custom handler so that when `log.fail` is invoked, we show a message and raise an exception.

Reply:
So I was using `@multiline` because we have logging levels also implemented. If we use:
```
log.debug('this is a multi-line output')
log.debug('    that can be put in multiple')
log.debug('    logging calls')
```
it will result in 
```
DEBUG: this is a multi-line output
DEBUG:    that can be put in multiple
DEBUG:    logging calls
```

Reply:
Having one DEBUG per line would be OK.